src/app/clusters/operational-state-server/OperationalStateDataProvider.cpp

```diff
@@ -0,0 +1,264 @@
+{
+    uint8_t buffer[kOperationalStateMaxSerializedSize];
+    MutableByteSpan bufferSpan(buffer);
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    ReturnErrorOnFailure(Load(DefaultStorageKeyAllocator::OperationalStateOpStateList(endpoint, clusterId).KeyName(), bufferSpan));
+
+    TLV::TLVReader reader;
+    TLV::TLVType outerType;
+
+    reader.Init(bufferSpan.data(), bufferSpan.size());
+    ReturnErrorOnFailure(reader.Next(TLV::TLVType::kTLVType_Array, TLV::AnonymousTag()));
+    ReturnErrorOnFailure(reader.EnterContainer(outerType));
+    while (reader.Next() != CHIP_ERROR_END_OF_TLV)
+    {
+        OperationalStateStructType operationalState;
+        ReturnErrorOnFailure(operationalState.Decode(reader));
+
+        if (targetOp.operationalStateID != operationalState.operationalStateID)
+            continue;
+
+        char * dest = const_cast<char *>(targetOp.OperationalStateLabel);
+        size_t len  = operationalState.operationalStateLabel.size();
+        memcpy(dest, operationalState.operationalStateLabel.data(), len);
+        targetOp.operationalStateLabel = CharSpan::fromCharString(targetOp.OperationalStateLabel);
+        break;
+    }
+
+    ReturnErrorOnFailure(reader.ExitContainer(outerType));
+
+    return err;
+}
+
+CHIP_ERROR OperationalStateDataProvider::StorePhaseList(EndpointId endpoint, ClusterId clusterId, const PhaseList & phaseList)
+{
+    uint8_t buffer[kOperationalStateMaxSerializedSize];
+    TLV::TLVWriter writer;
+    TLV::TLVType outerType;
+
+    writer.Init(buffer);
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Array, outerType));
+
+    for (auto const & phase : phaseList)
+    {
+        ReturnErrorOnFailure(writer.PutString(TLV::AnonymousTag(), phase.data()));
+    }
+    ReturnErrorOnFailure(writer.EndContainer(outerType));
+
+    return mPersistentStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::OperationalStatePhaseList(endpoint, clusterId).KeyName(),
+                                               buffer, static_cast<uint16_t>(writer.GetLengthWritten()));
+}
+
+CHIP_ERROR OperationalStateDataProvider::LoadPhaseList(EndpointId endpoint, ClusterId clusterId, PhaseListCharSpan ** phaseList,
+                                                       size_t & size)
+{
+    uint8_t buffer[kOperationalStateMaxSerializedSize];
+    MutableByteSpan bufferSpan(buffer);
+    size           = 0;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    ReturnErrorOnFailure(Load(DefaultStorageKeyAllocator::OperationalStatePhaseList(endpoint, clusterId).KeyName(), bufferSpan));
+
+    TLV::TLVReader reader;
+    TLV::TLVType outerType;
+
+    reader.Init(bufferSpan.data(), bufferSpan.size());
+    ReturnErrorOnFailure(reader.Next(TLV::TLVType::kTLVType_Array, TLV::AnonymousTag()));
+    ReturnErrorOnFailure(reader.EnterContainer(outerType));
+    size_t i                 = 0;
+    PhaseListCharSpan * head = nullptr;
+    while (reader.Next() != CHIP_ERROR_END_OF_TLV)
```

Comment:
If Next() returns an error that is NOT CHIP_ERROR_END_OF_TLV, then the code below will be working with garbage data.

Reply:
Done, Fix the bug.